boards: beagle: beagleconnect_freedom: Use sky13317 antenna switch

Ayush1325 opened the pull request:
- Use sky13317 device in dt instead of hack.
- Switch to using cc1352p1_launchxl board_antenna which also uses sky13317
- Fixes https://github.com/zephyrproject-rtos/zephyr/issues/71285

Ayush1325 committed:
drivers: misc: sky13317: Add sky13317 antenna

- sky13317 is used in cc1352p1_launchxl, beagleconnect_freedom and
  beagleplay_cc1352.
- Remove unused files
- Update maintainers

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

nordicjm commented on boards/beagle/beagleconnect_freedom/CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@
 zephyr_library()
 zephyr_library_sources(
-    board_antenna.c
+    ../../ti/cc1352p1_launchxl/board_antenna.c
```
would prefer if the file was copied over, having this is just asking for it to silently break if the vendor or board are renamed or if the board is removed

Ayush1325 replied:
Maybe it would be better to extract sky13317 to `drivers/misc`? This is used in cc1352p1_launchxl, beagleconnect_freedom and beagleplay_cc1352.

Ayush1325 replied:
Closing this now since sky13317 is now a separate driver.

Ayush1325 commented:
I'm not sure why the compliance check is failing here. Can anyone help?

vaishnavachath commented on boards/beagle/beagleconnect_freedom/Kconfig.beagleconnect_freedom:
```diff
@@ -6,3 +6,4 @@
 
 config BOARD_BEAGLECONNECT_FREEDOM
 	select SOC_CC1352P7
+	select SKY13317
```
Why do you need to do this if the default is y with depends on DT_ENABLED? and you it enabled in board DT also.

vaishnavachath commented on boards/ti/cc1352p1_launchxl/CMakeLists.txt:
```diff
@@ -1,4 +1,3 @@
 # SPDX-License-Identifier: Apache-2.0
 
 zephyr_library()
```
Shouldn't you be removing the file also if you plan to do this optimisation for this platform?

vaishnavachath commented on boards/ti/cc1352p1_launchxl/Kconfig.cc1352p1_launchxl:
```diff
@@ -5,3 +5,4 @@
 
 config BOARD_CC1352P1_LAUNCHXL
 	select SOC_CC1352P
+	select SKY13317
```
Do we really need to explicitly enable this?

vaishnavachath commented:
@Ayush1325 You will need to do a bit of cleanup while copying the driver, https://github.com/zephyrproject-rtos/zephyr/blob/83e8f1a2b0a75f719125f4c45f7a5d0c5f2f0594/drivers/misc/sky13317/sky13317.c#L52 _> drop this priority from board Kconfig.

nordicjm commented on boards/ti/cc1352p1_launchxl/cc1352p1_launchxl_defconfig:
```diff
@@ -22,3 +22,6 @@ CONFIG_SERIAL=y
 CONFIG_CONSOLE=y
 CONFIG_UART_CONSOLE=y
+
+# Antenna
+CONFIG_SKY13317_INIT_PRIO=70
```
Can be removed since the default is 70 anyway

nordicjm commented on boards/beagle/beagleconnect_freedom/beagleconnect_freedom_defconfig:
```diff
@@ -25,3 +25,6 @@ CONFIG_UART_CONSOLE=y
 # Adjust for oscillator capacitors
 CONFIG_CC13X2_CC26X2_XOSC_CAPARRAY_DELTA=0x02
+
+# Antenna
+CONFIG_SKY13317_INIT_PRIO=70
```
As above

nordicjm commented on drivers/misc/sky13317/sky13317.c:
```diff
@@ -49,7 +49,7 @@ const RFCC26XX_HWAttrsV2 RFCC26XX_hwAttrs = {
-					  POST_KERNEL, CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
+					  POST_KERNEL, CONFIG_SKY13317_INIT_PRIO, NULL);
 
 static const struct pinctrl_dev_config *ant_pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(0);
```
I have a problem with this driver, this is not a universal driver, no other platform can make use of this driver, the dts description lacks any sort of description for it and looking at the source file it essentially fits a device which has 2.4G output, PA output and sub-GHz output, so isn't at all generic and can't be used by any other board. @gmarull

Ayush1325 replied:
What should be added to the dts description? While the gpios is fixed (max 3), the pinctrl can be 0-8 depending on how the device is used.

I wonder if the driver can be changed to have nothing about the type of network it is configuring. Can you provide anything similar that can serve as an example. 
Alternatively, would it be fine to have three copies of the same `board_antenna` file?

gmarull reviewed (changes requested):
SKY13317 looks like a `single-pole, triple-throw (SP3T) antenna switch`, and this is creating a driver which can only be used with TI platforms. Definitely a -1, looks like this has been poorly architected.

Ayush1325 committed:
drivers: misc: sky13317: Make more generic

- Use seperate properties for v1, v2 and v3.
- Make pinctrl-0 required
- Gate cc1352 specific functionality (rf_callback) behind config
  features

Signed-off-by: Ayush Singh <ayushdevel1325@gmail.com>

Ayush1325 commented:
@gmarull @nordicjm I have moved `rf_callback` and related stuff behind config variable since that will be specific to soc networking capabilities. The init code should be generic. I have also made the bindings more specific by using dedicated v*-gpios for v1, v2 and v3 instead of bundling them together in gpios.

Does this seem better?

gmarull reviewed (changes requested):
not sure why I'm being asked to review again, when main concern hasn't been addressed

Ayush1325 commented:
> not sure why I'm being asked to review again, when main concern hasn't been addressed

Can you please clarify what you wish to be changed here. I have moved the TI specific stuff behind CC1352P and CC1352P7 config variables. However, since the `rf_callback` will be dependent on the networking capabilities of the device, I am not sure what more can be done for it. 

I think I can split the soc dependent code to `sky13317_cc13xx*.c` or something but are you maybe looking for something else? It does not seem like `sky13351` has any driver either (although the dt binding is present), so I am not sure what I should look towards as sample for what a Zephyr antenna switch should look like.

jadonk commented on drivers/misc/sky13317/sky13317.c:
```diff
@@ -79,9 +79,10 @@ int board_antenna_init(const struct device *dev)
  * Custom TI RFCC26XX callback for switching the on-board antenna mux on radio setup.
  */
-void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, void *arg)
+#if defined(CONFIG_SOC_CC1352P) || defined(CONFIG_SOC_CC1352P7)
```
I think we still need a SoC specific antenna driver, but it should be able to use the device-tree settings to operate the antenna. The role of the SoC-specific driver is to connect the calls that switch the antenna (assigned through device-tree) to the SoC-specific calls on mode-changes that might require antenna switches. The Kconfig for custom HWattrs should enable the usage of the CC13x board antenna driver.

jadonk commented on dts/bindings/misc/skyworks,sky13317.yaml:
```diff
@@ -0,0 +1,26 @@
+  v1-gpios:
+    type: phandle-array
+    required: true
+    description: v1 control pin
```
For me, the 'vX' description isn't very clear. Why 'v'?

I will say, are we overcomplicating this by taking the perspective of the antenna switch rather than the perspective of integrating the SoC's wireless framework and *possibly* needing to set GPIOs based on various modes?

jadonk reviewed (commented):
I tried to add some suggestions to make things more generic, but I believe what needs to happen is to create an SoC radio-mode driver that utilizes a board-specific device-tree to set various GPIOs if custom HWattrs are enabled in the config. The modes should drive the device-tree entries and the code should add the gpio setting if appropriate.

I'm probably not getting it, but I'm trying. :-D

Ayush1325 commented:
I have created another  PR which does not try to make a generic antenna switch bindings and just copies the current cc1352p1_launchxl antenna : #72332

I will close this PR in case that approach is preferable.

nordicjm commented on drivers/misc/sky13317/sky13317.c:
```diff
@@ -16,49 +16,49 @@
 #include <zephyr/drivers/pinctrl.h>
 
+#if defined(CONFIG_SOC_CC1352P) || defined(CONFIG_SOC_CC1352P7)
 #include <ti/drivers/rf/RF.h>
```
this is a no, as per the above comment

Ayush1325 replied:
Hi. Can you take a look at #72332 ?

nordicjm commented on drivers/misc/sky13317/sky13317.c:
```diff
@@ -16,49 +16,49 @@
 PINCTRL_DT_INST_DEFINE(0);
-DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL,
-					  POST_KERNEL, CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
+DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL, POST_KERNEL,
```
should use the for each dts syntax

Ayush1325 commented:
Closing in favor https://github.com/zephyrproject-rtos/zephyr/pull/72332
